src/backend/distributed/executor/merge_executor.c

```diff
@@ -0,0 +1,429 @@
+	 * partitions. Here we only lock targetRelation, we acquire necessary
+	 * locks on source tables during execution of those source queries.
+	 */
+	if (PartitionedTable(targetRelationId))
+	{
+		LockPartitionRelations(targetRelationId, RowExclusiveLock);
+	}
+
+	ereport(DEBUG1, (errmsg("Collect source query results on coordinator")));
+
+	List *prunedTaskList = NIL;
+	HTAB *shardStateHash =
+		ExecuteMergeSourcePlanIntoColocatedIntermediateResults(
+			targetRelationId,
+			mergeQuery,
+			sourceQuery->targetList,     /*TODO */
+			sourcePlan,
+			executorState,
+			intermediateResultIdPrefix);
+
+	ereport(DEBUG1, (errmsg("Create a MERGE task list that needs to be routed")));
+
+	/* generate tasks for the .. phase */
+	List *taskList =
+		GenerateTaskListWithColocatedIntermediateResults(targetRelationId, mergeQuery,
+														 intermediateResultIdPrefix);
+
+	/*
+	 * We cannot actually execute MERGE INTO ... tasks that read from
+	 * intermediate results that weren't created because no rows were
+	 * written to them. Prune those tasks out by only including tasks
+	 * on shards with connections.
+	 */
+	Task *task = NULL;
+	foreach_ptr(task, taskList)
+	{
+		uint64 shardId = task->anchorShardId;
+		bool shardModified = false;
+
+		hash_search(shardStateHash, &shardId, HASH_FIND, &shardModified);
+		if (shardModified)
+		{
+			prunedTaskList = lappend(prunedTaskList, task);
+		}
+	}
+
+	if (prunedTaskList == NIL)
+	{
+		/* No task to execute */
+		return;
+	}
+
+	ereport(DEBUG1, (errmsg("Execute MERGE task list")));
+	bool randomAccess = true;
+	bool interTransactions = false;
+	Assert(scanState->tuplestorestate == NULL);
+	scanState->tuplestorestate = tuplestore_begin_heap(randomAccess, interTransactions,
+													   work_mem);
+	TupleDesc tupleDescriptor = ScanStateGetTupleDescriptor(scanState);
+	TupleDestination *tupleDest =
+		CreateTupleStoreTupleDest(scanState->tuplestorestate, tupleDescriptor);
+	uint64 rowsMerged =
+		ExecuteTaskListIntoTupleDest(ROW_MODIFY_NONCOMMUTATIVE, prunedTaskList,
+									 tupleDest, hasReturning);
+	executorState->es_processed = rowsMerged;
+}
+
+
+/*
+ * sourceResultPartitionColumnIndex collects all Join conditions from the
+ * ON clause and verifies if there is a join, either left or right, with
+ * the distribution column of the given target. Once a match is found, it
+ * returns the index of that match in the source's target list.
+ */
+static int
+sourceResultPartitionColumnIndex(Query *mergeQuery, List *sourceTargetList,
+								 CitusTableCacheEntry *targetRelation)
+{
+	/* Get all the Join conditions from the ON clause */
+	List *mergeJoinConditionList = WhereClauseList(mergeQuery->jointree);
+
+	Var *targetColumn = targetRelation->partitionColumn;
+	bool joinedOnInsertColumn = false;
+	Var *sourceRepartitionVar = NULL;
+
+	OpExpr *validJoinClause =
+		SinglePartitionJoinClause(list_make1(targetColumn), mergeJoinConditionList);
+	if (!validJoinClause)
```

Comment:
Changed to

```
ERROR:  The required join operation is missing between the target's distribution column and any expression
 originating from the source. The issue may arise from either a non-equi-join or a mismatch in the datatypes 
of the columns being joined.
DETAIL:  Without a equi-join condition on the target's distribution column, the source rows cannot be 
efficiently redistributed, and the NOT-MATCHED condition cannot be evaluated unambiguously. This can
 result in incorrect or unexpected results when attempting to merge tables in a distributed setting
```